src/Servers/Kestrel/Core/src/Internal/Http/PathNormalizer.cs

```diff
@@ -50,193 +52,144 @@ public static string DecodePath(Span<byte> path, bool pathEncoded, string rawTar
+                    src[writtenLength++] = ByteSlash;
+                    return writtenLength;
+                }
+                else if (src[nextIndex] == ByteDot)
+                {
+                    // Remove the last segment and replace the path with /
+                    var lastSlashIndex = MemoryMarshal.CreateSpan(ref src[0], writtenLength).LastIndexOf(ByteSlash);
+
+                    // If this was the beginning of the string, then return /
+                    if (lastSlashIndex < 0)
                     {
-                        // C.  if the input buffer begins with a prefix of "/../" or "/..",
-                        //     where ".." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer and remove the last
-                        //     segment and its preceding "/" (if any) from the output
-                        //     buffer; otherwise,
-                        src += 3;
-
-                        if (dst > start)
-                        {
-                            do
-                            {
-                                dst--;
-                            } while (dst > start && *dst != ByteSlash);
-                        }
-
-                        continue;
+                        src[0] = ByteSlash;
+                        return 1;
                     }
-                    else if (ch2 == ByteDot && ch3 == ByteSlash)
+                    else
                     {
-                        // B.  if the input buffer begins with a prefix of "/./" or "/.",
-                        //     where "." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer; otherwise,
-                        src += 2;
-                        continue;
+                        writtenLength = lastSlashIndex + 1;
                     }
-
-                    break;
+                    return writtenLength;
+                }
+                else
+                {
+                    // Not a dot segment, copy the /. and bump the read pointer.
+                    slashDot.CopyTo(src[writtenLength..]);
+                    writtenLength += 2;
+                    readPointer = nextIndex;
+                }
             }
-
-            // E.  move the first path segment in the input buffer to the end of
-            //     the output buffer, including the initial "/" character (if
-            //     any) and any subsequent characters up to, but not including,
-            //     the next "/" character or the end of the input buffer.
-            do
+            // Ending with /.
+            else if (remainingLength == 2)
             {
-                *dst++ = ch1;
-                ch1 = *++src;
-            } while (src < end && ch1 != ByteSlash);
-        }
-
-        if (dst == start)
-        {
-            *dst++ = ByteSlash;
+                src[writtenLength++] = ByteSlash;
```

Comment:
Just increment `writtenLength`?

Reply:
There might be some other text already written at this point. The length check for 2 above is only for the remaining part of the text.